test/inductor/test_snode_runtime.py

```diff
@@ -165,6 +172,132 @@ def f(x):
         self.assertNotZero(calculate_runtime(f, *inp))
 
 
+class TestCommAnalysis(TestCase):
+    WORLD_SIZE: int = 8
+    RANKS = list(range(8))
+
+    def _verify_runtime_estimation(self, fn, inps):
+        store = FakeStore()
+        dist.init_process_group(
+            backend="fake", rank=0, world_size=self.WORLD_SIZE, store=store
+        )
+        try:
+            metrics.reset()
+            torch._dynamo.optimize(count_bytes_inductor)(fn)(*inps)
+            found_collective = False
+            for snode, runtime in metrics.node_runtimes:
+                if not is_collective(snode.node):
+                    continue
+                found_collective = True
+                # Inductor swallows errors from snode runtime estimations.
+                # We call estimate_nccl_collective_runtime in a white-box
+                # fashion here so potential issues can be surfaced in tests.
+                est = estimate_nccl_collective_runtime(snode.node)
+                self.assertNotZero(est)
+                # Also make sure estimate_nccl_collective_runtime works
+                # correctly in inductor.
+                self.assertNotZero(runtime)
+            # Make sure a collective kernel is found in graph
+            self.assertTrue(found_collective)
+        finally:
+            dist.destroy_process_group()
+
+    def test_legacy_all_reduce(self):
+        def fn(x):
+            r = c10d.all_reduce(x, "sum", "", self.RANKS, self.WORLD_SIZE)
+            return c10d.wait_tensor(r)
+
+        inp = T(10, 10)
+        self._verify_runtime_estimation(fn, (inp,))
+
+    def test_legacy_all_reduce_coalesced(self):
+        def fn(x):
+            rs = c10d.all_reduce_coalesced(x, "sum", "", self.RANKS, self.WORLD_SIZE)
+            return [c10d.wait_tensor(r) for r in rs]
+
+        inp = [T(10, 10), T(15, 15)]
+        self._verify_runtime_estimation(fn, (inp,))
+
+    def test_legacy_all_gather_into_tensor_coalesced(self):
+        def fn(x):
+            rs = c10d.all_gather_into_tensor_coalesced(
+                x,
+                "",
+                self.RANKS,
+                self.WORLD_SIZE,
+            )
+            return [c10d.wait_tensor(r) for r in rs]
+
+        inp = [T(10, 10), T(15, 15)]
+        self._verify_runtime_estimation(fn, (inp,))
+
+    def test_all_reduce(self):
+        def fn(x):
+            r = _c10d.all_reduce(x, "sum", "0")
+            return _c10d.wait_tensor(r)
+
+        inp = T(10, 10)
+        self._verify_runtime_estimation(fn, (inp,))
+
+    def test_all_reduce_coalesced(self):
+        def fn(x):
+            rs = _c10d.all_reduce_coalesced(x, "sum", "0")
+            return [_c10d.wait_tensor(r) for r in rs]
+
+        inp = [T(10, 10), T(15, 15)]
+        self._verify_runtime_estimation(fn, (inp,))
+
+    def test_all_gather_into_tensor(self):
+        def fn(x):
+            rs = _c10d.all_gather_into_tensor(
```

Comment:
hm, what's this doing- these are not functional collectives, so I wouldn't have expected them to work with wait_tensor.  or, is that counting on dynamo doing a rewrite?

Reply:
Hey @wconstab, these are dispatcher ops. For both the existing and new funcol, "tensor wrapping" happens inside dispatcher calls.

The reason I resorted to dispatcher ops for this test is that there's currently some issues with tracing some of the new funccols. This is okay for this test since it focuses on the IR. I will add tracing tests as I resolve the aforementioned issues.